policies/reputation: limit reputation score

gergosimonyi opened the pull request:
The goal of this is to prevent a brute force attack on a very high reputation (identifier, ip) tuple.

Note: reputation scores will stay the same when the tenant-wide limit is changed, so it is possible to temporarily have a score violating a newly set limit, until the score is first updated.

gergosimonyi committed:
add limits to reputation score

fheisler commented on authentik/tenants/models.py:
```diff
@@ -25,6 +25,8 @@
 
 DEFAULT_TOKEN_DURATION = "days=1"  # nosec
 DEFAULT_TOKEN_LENGTH = 60
+DEFAULT_REPUTATION_LOWER_LIMIT = -5
+DEFAULT_REPUTATION_UPPER_LIMIT = 5
```
How does it get used in any default templates? 5 sounds a lot more reasonable than 1k to me, otherwise it mostly recreates the issue we're trying to avoid here

rissson replied:
The default threshold in the reputation policy is -5.

gergosimonyi replied:
I believe these are good defaults, yes. Upper limit is to prevent an attacker from spamming "too much". Lower limit is to require 1 successful login with captcha with the example policy to no longer require a captcha.

gergosimonyi commented on authentik/tenants/models.py:
```diff
@@ -70,6 +72,14 @@ class Tenant(TenantMixin, SerializerModel):
             "Events will be deleted after this duration.(Format: weeks=3;days=2;hours=3,seconds=2)."
         ),
     )
+    reputation_lower_limit = models.IntegerField(
+        help_text=_("Reputation cannot decrease lower than this value."),
+        default=DEFAULT_REPUTATION_LOWER_LIMIT,
+    )
+    reputation_upper_limit = models.IntegerField(
+        help_text=_("Reputation cannot increase higher than this value."),
+        default=DEFAULT_REPUTATION_UPPER_LIMIT,
+    )
```
Yeah, I didn't want to limit admins unreasonably, but I went too far. This seems reasonable and simplifies much.

gergosimonyi committed:
limit reputation score limits

Upper to non-negative, Lower to non-positive

gergosimonyi committed:
simplify tests

gergosimonyi committed:
Merge branch 'main' into policies/reputation/limit-reputaiton-score

gergosimonyi commented on authentik/policies/reputation/tests.py:
```diff
@@ -17,36 +20,47 @@ class TestReputationPolicy(TestCase):
+        self.password = make_password("password")
         # We need a user for the one-to-one in userreputation
-        self.user = User.objects.create(username=self.test_username)
+        self.user = User.objects.create(username=self.username, password=self.password)
```
That's simpler. Thanks!

gergosimonyi committed:
"fix" bandit false-positives

gergosimonyi committed:
move magic numbers to constants

Is it too much to ask for a world in which I can just import these
straight from Python?
